<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Fast Estimation of Gaussian Mixture Copula Models • GMCM</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Fast Estimation of Gaussian Mixture Copula Models">
<meta property="og:description" content="Unsupervised Clustering and Meta-analysis using Gaussian Mixture
    Copula Models.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">GMCM</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/AEBilgrau/GMCM">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">


<div id="gmcm" class="section level2">
<h2 class="hasAnchor">
<a href="#gmcm" class="anchor"></a>GMCM</h2>
<div id="fast-estimation-of-gaussian-mixture-copula-models" class="section level3">
<h3 class="hasAnchor">
<a href="#fast-estimation-of-gaussian-mixture-copula-models" class="anchor"></a>Fast estimation of Gaussian Mixture Copula Models</h3>
<p>The <a href="https://cran.r-project.org/package=GMCM"><strong>GMCM</strong> package</a> offers R functions that perform high-dimensional meta-analysis <a href="http://arxiv.org/pdf/1110.4705.pdf">(Li et. al., 2011)</a> and general unsupervised cluster analysis <a href="http://ieeexplore.ieee.org/xpl/articleDetails.jsp?arnumber=6137392">(Tewari et. al., 2011)</a> using Gaussian Copula Mixture Models in a very fast manner [<a href="https://www.jstatsoft.org/article/view/v070i02">1</a>].</p>
<p>Gaussian copula mixture models (GMCMs) are a very flexible alternative to Gaussian mixture models in unsupervised cluster analysis for continuous data where non-Gaussian clusters are present. GMCMs model the ranks of the observed data and are thus invariant to monotone increasing transformations of the data, i.e. they are semi-parametric and only the ordering of the data is important. Alternatively, a special-case of the GMCMs can be used for a novel meta-analysis approach in high-dimensional settings. In this context, the model tries to cluster results which agree and do not agree on statistical evidence into a reproducible and irreproducible group.</p>
<p>The optimization of the complicated likelihood function is difficult, however. <strong>GMCM</strong> utilizes <a href="https://github.com/RcppCore/Rcpp"><strong>Rcpp</strong></a> and <a href="https://github.com/RcppCore/RcppArmadillo"><strong>RcppArmadillo</strong></a> to evaluate the likelihood function quickly and arrive at a parameter estimate using either standard numerical optimization routines or an pseudo EM algorithm.</p>
<p>Additional information, documentation, help, and examples can be found by running <code><a href="https://rdrr.io/r/utils/help.html">help(package = "GMCM")</a></code> or <code><a href="reference/GMCM-package.html">?GMCM</a></code> in <strong>R</strong>. The paper [1] is also found as a vignette by <code>vignette("GMCM-JStatSoft")</code>. The core user functions of <strong>GMCM</strong> are <code>fit.full.GMCM</code> and <code>fit.meta.GMCM</code>.</p>
</div>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>The released and tested version of <strong>GMCM</strong> is available at <a href="https://cran.r-project.org/package=GMCM">CRAN</a> (Comprehensive R Archive Network). It can be installed from within R by running</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"GMCM"</span>)</a></code></pre></div>
<p>If you wish to install the latest version of <strong>GMCM</strong> directly from the master branch at GitHub, run</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co">#install.packages("remotes")  # Install remotes if needed</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2">remotes<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/remotes/man/install_github.html">install_github</a></span>(<span class="st">"AEBilgrau/GMCM"</span>)</a></code></pre></div>
<p>Note, that this version is in development and is likely different from the version at CRAN. As such, it may be unstable. Be sure that you have the <a href="https://support.rstudio.com/hc/en-us/articles/200486498-Package-Development-Prerequisites">package development prerequisites</a> if you wish to install the package from the source.</p>
<p>When installed, run <code><a href="reference/runGMCM.html">GMCM::runGMCM()</a></code> to launch a local instance of the GMCM shiny application also available online <a href="https://gmcm.shinyapps.io/GMCM/">shinyapps.io</a>). Run <code><a href="https://rdrr.io/r/utils/news.html">news(package = "GMCM")</a></code> to view the latest changes of GMCM.</p>
<p>For previous versions of <strong>GMCM</strong>, visit the old <a href="https://github.com/AEBilgrau/GMCM/releases">releases at GitHub</a> or the <a href="https://cran.r-project.org/src/contrib/Archive/GMCM/">archive at CRAN.</a></p>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<ol>
<li>Anders Ellern Bilgrau, Poul Svante Eriksen, Jakob Gulddahl Rasmussen, Hans Erik Johnsen, Karen Dybkaer, Martin Boegsted (2016). <strong>GMCM: Unsupervised Clustering and Meta-Analysis Using Gaussian Mixture Copula Models.</strong> Journal of Statistical Software, 70(2), 1-23. <a href="https://www.jstatsoft.org/article/view/v070i02">doi:10.18637/jss.v070.i02</a>
</li>
</ol>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<div id="shiny-application" class="section level3">
<h3 class="hasAnchor">
<a href="#shiny-application" class="anchor"></a>Shiny application</h3>
<p>A <a href="https://shiny.rstudio.com"><strong>shiny</strong></a> application is available online <a href="https://gmcm.shinyapps.io/GMCM/">shinyapps.io</a>. To start local instance of the application, make sure <strong>GMCM</strong> is installed and run</p>
<p><code>{r, eval=FALSE} GMCM::runGMCM()</code></p>
<p>The application suppots both the general GMCM for unsupervised clustering as well as the special GMCM for meta analysis.</p>
</div>
<div id="meta-analysis-example" class="section level3">
<h3 class="hasAnchor">
<a href="#meta-analysis-example" class="anchor"></a>Meta Analysis example</h3>
<p>This is a very short tutorial for using the special GMCM for meta analysis. To illustrate we load the <code>u133VsExon</code> dataset within the package. The dataset contains 19,577 <em>p</em>-values for the null hypothesis of no differential gene expression between two cell types for each of two different experiments called <code>u133</code> and <code>exon</code>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co"># Load and show data</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(u133VsExon)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(u133VsExon, <span class="dt">n =</span> <span class="dv">3</span>)</a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="co">#                        u133         exon</span></a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="co">#ENSG00000265096 1.756104e-01 1.072572e-01</span></a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="co">#ENSG00000152495 1.779757e-03 6.741108e-10</span></a>
<a class="sourceLine" id="cb3-7" data-line-number="7"><span class="co">#ENSG00000198040 5.370574e-03 1.505019e-03</span></a></code></pre></div>
<p>See e.g. <code><a href="reference/u133VsExon.html">?u133VsExon</a></code> for more information. Next, we subset the data, rank it, and visualize it:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># Subsetting data to reduce computation time</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">u133VsExon &lt;-<span class="st"> </span>u133VsExon[<span class="dv">1</span><span class="op">:</span><span class="dv">5000</span>, ]</a>
<a class="sourceLine" id="cb4-3" data-line-number="3"></a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="co"># Ranking and scaling, remember large values should be critical to the null!</span></a>
<a class="sourceLine" id="cb4-5" data-line-number="5">uhat &lt;-<span class="st"> </span><span class="kw"><a href="reference/Uhat.html">Uhat</a></span>(<span class="dv">1</span> <span class="op">-</span><span class="st"> </span>u133VsExon)</a>
<a class="sourceLine" id="cb4-6" data-line-number="6"></a>
<a class="sourceLine" id="cb4-7" data-line-number="7"><span class="kw"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="dt">mfrow =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>,<span class="dv">2</span>))  <span class="co"># Visualizing P-values and the ranked P-values</span></a>
<a class="sourceLine" id="cb4-8" data-line-number="8"><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(u133VsExon, <span class="dt">cex =</span> <span class="fl">0.5</span>, <span class="dt">pch =</span> <span class="dv">4</span>, <span class="dt">col =</span> <span class="st">"tomato"</span>, <span class="dt">main =</span> <span class="st">"P-values"</span>,</a>
<a class="sourceLine" id="cb4-9" data-line-number="9">     <span class="dt">xlab =</span> <span class="st">"P-value (U133)"</span>, <span class="dt">ylab =</span> <span class="st">"P-value (Exon)"</span>)</a>
<a class="sourceLine" id="cb4-10" data-line-number="10"><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(uhat, <span class="dt">cex =</span> <span class="fl">0.5</span>, <span class="dt">pch =</span> <span class="dv">4</span>, <span class="dt">col =</span> <span class="st">"tomato"</span>, <span class="dt">main =</span> <span class="st">"Ranked P-values"</span>,</a>
<a class="sourceLine" id="cb4-11" data-line-number="11">     <span class="dt">xlab =</span> <span class="st">"rank(1-P) (U133)"</span>, <span class="dt">ylab =</span> <span class="st">"rank(1-P) (Exon)"</span>)</a></code></pre></div>
<p>Here each point represent a gene. The genes in the lower left of the first panel and correspondingly in the upper right of the second panel are the seemingly reproducible genes. They have a low <em>p</em>-value and thus a high rank and <em>both</em> experiments. Next, we do the actual fit using the core user function <code>fit.meta.GMCM</code> with a L-BFGS-like procedure and subsequently compute the IDR values:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">fit &lt;-<span class="st"> </span><span class="kw"><a href="reference/fit.meta.GMCM.html">fit.meta.GMCM</a></span>(uhat, <span class="dt">init.par =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.5</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="fl">0.5</span>), </a>
<a class="sourceLine" id="cb5-2" data-line-number="2">                     <span class="dt">method =</span> <span class="st">"L-BFGS"</span>, <span class="dt">pgtol =</span> <span class="fl">1e-2</span>, <span class="dt">verbose =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb5-3" data-line-number="3">                     </a>
<a class="sourceLine" id="cb5-4" data-line-number="4">idr &lt;-<span class="st"> </span><span class="kw"><a href="reference/get.IDR.html">get.IDR</a></span>(uhat, <span class="dt">par =</span> fit)  <span class="co"># Compute IDR values and classify</span></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="kw"><a href="https://rdrr.io/r/base/table.html">table</a></span>(idr<span class="op">$</span>K)  <span class="co"># Show results, 1 = irreproducible, 2 = reproducible</span></a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="co">#   1    2 </span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="co">#4136  864 </span></a></code></pre></div>
<p>Where we see that 864 of the 5000 genes are deemed reproducible by the model at the default 0.05 threshold. The clustering results and the reproducible genes can be visualized in the following manner:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="dt">mfrow =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>,<span class="dv">2</span>))</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(u133VsExon, <span class="dt">cex =</span> <span class="fl">0.5</span>, <span class="dt">pch =</span> <span class="dv">4</span>, <span class="dt">main =</span> <span class="st">"Classified genes"</span>,</a>
<a class="sourceLine" id="cb6-3" data-line-number="3">     <span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"tomato"</span>, <span class="st">"steelblue"</span>)[idr<span class="op">$</span>K],</a>
<a class="sourceLine" id="cb6-4" data-line-number="4">     <span class="dt">xlab =</span> <span class="st">"P-value (U133)"</span>, <span class="dt">ylab =</span> <span class="st">"P-value (Exon)"</span>)</a>
<a class="sourceLine" id="cb6-5" data-line-number="5"><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(uhat, <span class="dt">cex =</span> <span class="fl">0.5</span>, <span class="dt">pch =</span> <span class="dv">4</span>, <span class="dt">main =</span> <span class="st">"Classified genes"</span>,</a>
<a class="sourceLine" id="cb6-6" data-line-number="6">     <span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"tomato"</span>, <span class="st">"steelblue"</span>)[idr<span class="op">$</span>K],</a>
<a class="sourceLine" id="cb6-7" data-line-number="7">     <span class="dt">xlab =</span> <span class="st">"rank(1-P) (U133)"</span>, <span class="dt">ylab =</span> <span class="st">"rank(1-P) (Exon)"</span>)</a></code></pre></div>
<p>The model indeed capture genes in the upper right (of the second panel) and classify them as reproducible.</p>
<hr>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from CRAN at <br><a href="https://cloud.r-project.org/package=GMCM">https://​cloud.r-project.org/​package=GMCM</a>
</li>
<li>Browse source code at <br><a href="https://github.com/AEBilgrau/GMCM">https://​github.com/​AEBilgrau/​GMCM</a>
</li>
<li>Report a bug at <br><a href="https://github.com/AEBilgrau/GMCM/issues">https://​github.com/​AEBilgrau/​GMCM/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>GPL (&gt;=2)</li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing GMCM</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Anders Ellern Bilgrau <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/AEBilgrau/GMCM"><img src="https://api.travis-ci.org/AEBilgrau/GMCM.svg?branch=master" alt="Build Status"></a></li>
<li><a href="https://cran.r-project.org/package=GMCM"><img src="http://www.r-pkg.org/badges/version/GMCM" alt="CRAN version"></a></li>
<li><a href="https://cran.r-project.org/package=GMCM"><img src="http://cranlogs.r-pkg.org/badges/GMCM" alt="CRAN RStudio mirror downloads"></a></li>
<li><a href="https://cran.r-project.org/package=GMCM"><img src="http://cranlogs.r-pkg.org/badges/grand-total/GMCM" alt="Total CRAN downloads"></a></li>
<li><a href="https://coveralls.io/github/AEBilgrau/GMCM?branch=master"><img src="https://coveralls.io/repos/github/AEBilgrau/GMCM/badge.svg?branch=master" alt="Coverage Status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Anders Ellern Bilgrau.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.9900.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
