<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Meta analysis with special GMCMs • GMCM</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Meta analysis with special GMCMs">
<meta property="og:description" content="GMCM">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-155943904-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-155943904-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GMCM</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/AEBilgrau/GMCM/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Meta analysis with special GMCMs</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/AEBilgrau/GMCM/blob/master/vignettes/usage-example-special-model.Rmd"><code>vignettes/usage-example-special-model.Rmd</code></a></small>
      <div class="hidden name"><code>usage-example-special-model.Rmd</code></div>

    </div>

    
    
<p>This is a quick tutorial for using the special GMCM for meta analysis.</p>
<div id="initialization" class="section level2">
<h2 class="hasAnchor">
<a href="#initialization" class="anchor"></a>Initialization</h2>
<p>The <strong>GMCM</strong><sup><a href="http://doi.org/10.18637/jss.v070.i02">1</a></sup> package is loaded.</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="co">#install.packages("GMCM")  # Uncomment to install the GMCM package</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st"><a href="http://AEBilgrau.github.io/GMCM">"GMCM"</a></span>)</pre></div>
<p>If <strong>GMCM</strong> is <em>not</em> installed, please uncomment the above line and rerun the script.</p>
</div>
<div id="loading-data" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-data" class="anchor"></a>Loading data</h2>
<p>The data is loaded and the first rows are printed. To illustrate we load the <code>u133VsExon</code> dataset within the package. The dataset contains 19,577 <em>p</em>-values for the null hypothesis of no differential gene expression between two cell types for each of two different experiments called <code>u133</code> and <code>exon</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="kw">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/get.html">get</a></span>(<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"u133VsExon"</span>))
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">x</span>, n = <span class="fl">4</span>)</pre></div>
<pre><code>##                         u133         exon
## ENSG00000265096 0.1756103672 1.072572e-01
## ENSG00000152495 0.0017797571 6.741108e-10
## ENSG00000198040 0.0053705738 1.505019e-03
## ENSG00000229092 0.0006693361 6.755118e-05</code></pre>
<p>See <code><a href="../reference/u133VsExon.html">help("u133VsExon")</a></code> for more information.</p>
<p>Next, we subset the data to simply reduce computation time. After that, we will rank it, and visualize it.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="kw">x</span> <span class="op">&lt;-</span> <span class="kw">x</span>[<span class="fl">1</span><span class="op">:</span><span class="fl">5000</span>, ]</pre></div>
<p>The values above are p-values where <em>small</em> values indicate strong evidence—contrary to what is expected by the special model. In the special model, <em>large</em> values should be critical to the null hypothesis.</p>
<p>Therefore we ranks and scale 1 - p:</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="kw">u</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Uhat.html">Uhat</a></span>(<span class="fl">1</span> <span class="op">-</span> <span class="kw">x</span>)</pre></div>
<p>The original and ranked data is plotted:</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(mfrow = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">2</span>))  <span class="co"># Visualizing P-values and the ranked P-values</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">x</span>, cex = <span class="fl">0.5</span>, pch = <span class="fl">4</span>, col = <span class="st">"tomato"</span>, main = <span class="st">"P-values"</span>,
     xlab = <span class="st">"P-value (U133)"</span>, ylab = <span class="st">"P-value (Exon)"</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">u</span>, cex = <span class="fl">0.5</span>, pch = <span class="fl">4</span>, col = <span class="st">"tomato"</span>, main = <span class="st">"Ranked P-values"</span>,
     xlab = <span class="st">"rank(1-P) (U133)"</span>, ylab = <span class="st">"rank(1-P) (Exon)"</span>)</pre></div>
<p><img src="usage-example-special-model_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<p>Here each point represent a gene. The genes in the lower left of the first panel and correspondingly in the upper right of the second panel are the seemingly <em>reproducible</em> genes. They have a low <em>p</em>-value and thus a high rank in <em>both</em> experiments. The genes in the upper left and lower right are the ones that are apparently spurious results; they have a low <em>p</em>-value in only one experiment.</p>
</div>
<div id="initial-parameters" class="section level2">
<h2 class="hasAnchor">
<a href="#initial-parameters" class="anchor"></a>Initial parameters</h2>
<p>The initial parameters are set to</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="kw">init_par</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(pie1 = <span class="fl">0.6</span>, mu = <span class="fl">1</span>, sigma = <span class="fl">1</span>, rho = <span class="fl">0.2</span>)</pre></div>
</div>
<div id="model-fitting" class="section level2">
<h2 class="hasAnchor">
<a href="#model-fitting" class="anchor"></a>Model fitting</h2>
<p>With the data loaded and prepared, the model is can be fitted with the defined initial parameters.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="kw">par</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fit.meta.GMCM.html">fit.meta.GMCM</a></span>(u = <span class="kw">u</span>,
                     init.par = <span class="kw">init_par</span>,
                     method = <span class="st">"NM"</span>,
                     max.ite = <span class="fl">1000</span>,
                     verbose = <span class="fl">FALSE</span>)
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="kw">par</span>)</pre></div>
<pre><code>##      pie1        mu     sigma       rho 
## 0.7489036 1.9494403 1.1066673 0.7360319</code></pre>
<p>We here use the Nelder-Mead fitting method with with a maximum number of iterations of <code>1000</code>.</p>
</div>
<div id="meta-analysis-with-unsupervised-clustering" class="section level2">
<h2 class="hasAnchor">
<a href="#meta-analysis-with-unsupervised-clustering" class="anchor"></a>Meta analysis with unsupervised clustering</h2>
<p>The estimated parameters are used to calculate the local and adjusted irreproducibility discovery rates:</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="kw">meta_IDR_thres</span> <span class="op">&lt;-</span> <span class="fl">0.05</span>
<span class="kw">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get.IDR.html">get.IDR</a></span>(<span class="kw">x</span>, par = <span class="kw">par</span>,
               threshold = <span class="kw">meta_IDR_thres</span>) <span class="co"># Compute IDR</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span>(<span class="kw">out</span>)</pre></div>
<pre><code>## List of 5
##  $ idr      : num [1:5000] 0.872 0.99 0.976 0.981 0.203 ...
##  $ IDR      : num [1:5000] 0.4894 0.769 0.7032 0.7252 0.0645 ...
##  $ l        : int 431
##  $ threshold: num 0.05
##  $ Khat     : num [1:5000] 1 1 1 1 1 1 1 1 1 1 ...</code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="kw">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get.IDR.html">get.IDR</a></span>(<span class="kw">u</span>, par = <span class="kw">par</span>, threshold = <span class="kw">meta_IDR_thres</span>)
<span class="kw">below</span> <span class="op">&lt;-</span> <span class="kw">out</span>[[<span class="st">"IDR"</span>]] <span class="op">&lt;</span> <span class="kw">meta_IDR_thres</span>
<span class="kw">out</span><span class="op">$</span><span class="kw">l</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="kw">below</span>)
<span class="kw">out</span><span class="op">$</span><span class="kw">Khat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="kw">below</span>, <span class="fl">2</span>, <span class="fl">1</span>)</pre></div>
<p>By default, <code>get.IDR</code> computes thresholds on the adjusted IDR. The local irreproducibility discovery rate correspond to the posterior probability of the point originating from the irreproducible component.</p>
</div>
<div id="results" class="section level2">
<h2 class="hasAnchor">
<a href="#results" class="anchor"></a>Results</h2>
<p>The classes are counted by</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="kw">out</span><span class="op">$</span><span class="kw">Khat</span>)</pre></div>
<pre><code>## 
##    1    2 
## 4136  864</code></pre>
<p>Where we see how many of the 5000 genes are deemed reproducible and irreproducible.</p>
<p>The results are also displayed by plotting</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">x</span>, col = <span class="kw">out</span><span class="op">$</span><span class="kw">Khat</span>, asp = <span class="fl">1</span>) <span class="co"># Plot of raw values</span></pre></div>
<p><img src="usage-example-special-model_files/figure-html/plot_results-1.png" width="700"></p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">u</span>, col = <span class="kw">out</span><span class="op">$</span><span class="kw">Khat</span>, asp = <span class="fl">1</span>) <span class="co"># Plot of copula values</span></pre></div>
<p><img src="usage-example-special-model_files/figure-html/plot_results-2.png" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="kw">z</span> <span class="op">&lt;-</span> <span class="kw">GMCM</span>:::<span class="fu"><a href="../reference/dgmcm.loglik.html">qgmm.marginal</a></span>(<span class="kw">u</span>, theta = <span class="fu"><a href="../reference/full2meta.html">meta2full</a></span>(<span class="kw">par</span>, d = <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="kw">u</span>))) <span class="co"># Estimate latent process</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">z</span>, col = <span class="kw">out</span><span class="op">$</span><span class="kw">Khat</span>, asp = <span class="fl">1</span>) <span class="co"># Plot of estimated latent process</span></pre></div>
<p><img src="usage-example-special-model_files/figure-html/plot_results-3.png" width="700"></p>
<p>The model indeed capture genes in the upper right and classify them as reproducible.</p>
<p>The fitted <code>par</code> object can be converted to a <code>theta</code> object which can be plotted directly:</p>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="fu"><a href="../reference/full2meta.html">meta2full</a></span>(<span class="kw">par</span>, d = <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="kw">u</span>)))</pre></div>
<p><img src="usage-example-special-model_files/figure-html/plot_theta-1.png" width="700"></p>
<div id="session-information" class="section level3">
<h3 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h3>
<p>This report was generated using <strong>rmarkdown</strong><sup><a href="https://bookdown.org/yihui/rmarkdown/parameterized-reports.html">2</a></sup> and <strong>knitr</strong><sup><a href="https://yihui.name/knitr/demo/stitch/">3</a></sup> under the session given below. The report utilizes <a href="https://bookdown.org/yihui/rmarkdown/parameterized-reports.html">parameterized reports</a> and <a href="https://yihui.name/knitr/demo/stitch/"><code>knitr::spin</code></a>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</pre></div>
<pre><code>## R version 4.0.2 (2020-06-22)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 16.04.6 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/openblas-base/libblas.so.3
## LAPACK: /usr/lib/libopenblasp-r0.2.18.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] GMCM_1.4.1 knitr_1.29
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.5         digest_0.6.25      crayon_1.3.4       rprojroot_1.3-2   
##  [5] assertthat_0.2.1   R6_2.4.1           backports_1.1.10   magrittr_1.5      
##  [9] ellipse_0.4.2      evaluate_0.14      stringi_1.5.3      rlang_0.4.7       
## [13] fs_1.5.0           ragg_0.3.1         rmarkdown_2.3      pkgdown_1.6.1.9000
## [17] desc_1.2.0         tools_4.0.2        stringr_1.4.0      yaml_2.2.1        
## [21] xfun_0.17          compiler_4.0.2     systemfonts_0.3.1  cpp11_0.2.1       
## [25] memoise_1.1.0      htmltools_0.5.0</code></pre>
</div>
<div id="references" class="section level3">
<h3 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h3>
<p>Please cite the <strong>GMCM</strong> paper<sup><a href="http://doi.org/10.18637/jss.v070.i02">1</a></sup> if you use the package or shiny app.</p>
<ol style="list-style-type: decimal">
<li>Bilgrau AE, Eriksen PS, Rasmussen JG, Johnsen HE, Dybkaer K, Boegsted M (2016). “GMCM: Unsupervised Clustering and Meta-Analysis Using Gaussian Mixture Copula Models.” <em>Journal of Statistical Software</em>, <em>70</em>(2), 1-23. doi: 10.18637/jss.v070.i02 (URL: <a href="https://doi.org/10.18637/jss.v070.i02" class="uri">https://doi.org/10.18637/jss.v070.i02</a>).</li>
<li>Xie Y (2020). <em>knitr: A General-Purpose Package for Dynamic Report Generation in R</em>. R package version 1.29, &lt;URL: <a href="https://yihui.org/knitr/" class="uri">https://yihui.org/knitr/</a>&gt;.</li>
<li>Xie Y (2015). <em>Dynamic Documents with R and knitr</em>, 2nd edition. Chapman and Hall/CRC, Boca Raton, Florida. ISBN 978-1498716963, &lt;URL: <a href="https://yihui.org/knitr/" class="uri">https://yihui.org/knitr/</a>&gt;.</li>
<li>Xie Y (2014). “knitr: A Comprehensive Tool for Reproducible Research in R.” In Stodden V, Leisch F, Peng RD (eds.), <em>Implementing Reproducible Computational Research</em>. Chapman and Hall/CRC. ISBN 978-1466561595, &lt;URL: <a href="http://www.crcpress.com/product/isbn/9781466561595" class="uri">http://www.crcpress.com/product/isbn/9781466561595</a>&gt;.</li>
<li>Allaire J, Xie Y, McPherson J, Luraschi J, Ushey K, Atkins A, Wickham H, Cheng J, Chang W, Iannone R (2020). <em>rmarkdown: Dynamic Documents for R</em>. R package version 2.3, &lt;URL: <a href="https://github.com/rstudio/rmarkdown" class="uri">https://github.com/rstudio/rmarkdown</a>&gt;.</li>
<li>Xie Y, Allaire J, Grolemund G (2018). <em>R Markdown: The Definitive Guide</em>. Chapman and Hall/CRC, Boca Raton, Florida. ISBN 9781138359338, &lt;URL: <a href="https://bookdown.org/yihui/rmarkdown" class="uri">https://bookdown.org/yihui/rmarkdown</a>&gt;.</li>
</ol>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Anders Ellern Bilgrau.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
